lib/Service/Appointments/BookingCalendarWriter.php

```diff
@@ -0,0 +1,142 @@
+ * version 3 of the License, or any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU AFFERO GENERAL PUBLIC LICENSE for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public
+ * License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Calendar\Service\Appointments;
+
+use DateTime;
+use DateTimeImmutable;
+use OCA\Calendar\Db\AppointmentConfig;
+use OCA\DAV\CalDAV\CalendarImpl;
+use OCP\Calendar\Exceptions\CalendarException;
+use OCP\Calendar\ICreateFromString;
+use OCP\Calendar\IManager;
+use OCP\IConfig;
+use OCP\IUserManager;
+use OCP\Security\ISecureRandom;
+use RuntimeException;
+use Sabre\VObject\Component\VCalendar;
+
+class BookingCalendarWriter {
+
+	/** @var IConfig */
+	private $config;
+
+	/** @var IManager */
+	private $manager;
+
+	/** @var IUserManager */
+	private $userManager;
+
+	/** @var ISecureRandom */
+	private $random;
+
+	public function __construct(IConfig $config,
+								IManager $manager,
+								IUserManager $userManager,
+								ISecureRandom $random) {
+		$this->config = $config;
+		$this->manager = $manager;
+		$this->userManager = $userManager;
+		$this->random = $random;
+	}
+
+	/**
+	 * @param AppointmentConfig $config
+	 * @param DateTimeImmutable $start
+	 * @param string $name
+	 * @param string $email
+	 * @param string $description
+	 *
+	 * @throws RuntimeException
+	 *
+	 */
+	public function write(AppointmentConfig $config, DateTimeImmutable $start, string $displayName, string $email, ?string $description = null) : void {
+		$calendar = current($this->manager->getCalendarsForPrincipal($config->getPrincipalUri(), [$config->getTargetCalendarUri()]));
+		if (!$calendar || !($calendar instanceof ICreateFromString)) {
+			throw new RuntimeException('Could not find a public writable calendar for this principal');
+		}
+
+		$organizer = $this->userManager->get($config->getUserId());
+		if ($organizer === null) {
+			throw new RuntimeException('Organizer not registered user for this instance');
+		}
+
+		$vcalendar = new VCalendar([
+			'CALSCALE' => 'GREGORIAN',
+			'VERSION' => '2.0',
+			'VEVENT' => [
+				'SUMMARY' => $config->getName(),
+				'STATUS' => 'CONFRIMED',
+				'DTSTART' => $start,
+				'DTEND' => $start->setTimestamp($start->getTimestamp() + ($config->getLength() * 60)) // *60 can be removed as soon as the switch to seconds as intervals etc is complete
+			]
+		]);
+
+		if($description !== null) {
+			$vcalendar->VEVENT->add('DESCRIPTION', $description);
+		}
+		$vcalendar->VEVENT->add(
+			'ORGANIZER',
+			'mailto:' . $organizer->getEMailAddress(),
+			[
+				'CN' => $organizer->getDisplayName(),
+				'CUTYPE' => 'INDIVIDUAL',
+				'PARTSTAT' => 'ACCEPTED'
+			]
+		);
+		$vcalendar->VEVENT->add(
```

Comment:
Can't remember why the organizer has to be added twice as an `ORGANIZER` and as an `ATTENDEE`. For email notifications ?

Reply:
because the organizer is not automatically also an attendee